### PR TITLE
[24575] Instrument timeline visibility state through query

### PR DESCRIPTION
**Pull request by oliverguenther**
Implements part of https://community.openproject.com/work_packages/24575

### CODE REVIEW

File: frontend/app/helpers/angular-rx-utils.ts [hunk 1]
```diff
@@ -42,10 +42,10 @@ export function asyncTest<T>(done: (error?: any) => void, fn: (value: T) => any)
 }
 
 export function scopeDestroyed$(scope: IScope): Observable<IScope> {
-  return Observable.create((s: any) => {
+  return Observable.create((s:Observer<IScope>) => {
     scope.$on("$destroy", () => {
-      s.onNext(scope);
-      s.onCompleted();
+      s.next(scope);
```
Comment (oliverguenther) [inline frontend/app/helpers/angular-rx-utils.ts:47]: /cc @romanroe This caused errors when moving between (UI router) states due to the `scopedDestroyed$` raising errors.

File: frontend/app/components/api/api-v3/hal-resource-dms/payload-dm.service.ts [hunk 2]
```diff
@@ -58,7 +58,7 @@ export class PayloadDmService {
     }
 
     _.each(nonLinkProperties, property => {
-      if (resource[property]) {
+      if (resource.hasOwnProperty(property)) {
```
Comment (oliverguenther) [inline frontend/app/components/api/api-v3/hal-resource-dms/payload-dm.service.ts:61]: /cc @ulferts This caused saving a query with any `false` value (e.g., `displaySums: false` or, in this case, timelineVisible) to ignore that value.

File: db/migrate/20170404110156_extend_query_model.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class ExtendQueryModel < ActiveRecord::Migration[5.0]
+  def change
+    add_column :queries, :timeline_visible, :boolean, default: false
```
Comment (ulferts) [inline db/migrate/20170404110156_extend_query_model.rb:3]: Just a matter of taste but I'd prefer it to be called simply `timeline`

File: frontend/app/components/api/api-v3/hal-resource-dms/payload-dm.service.ts (hunk 2 above)
Comment (ulferts) [reply frontend/app/components/api/api-v3/hal-resource-dms/payload-dm.service.ts:61]: Thanks

**Review by ulferts — CHANGES_REQUESTED** [review]
Added functionality looks good. No Bugs detected. 

What is missing in my opinion:
* The timeline state is not reflected in the url. Therefore, the user can not send the url and expect the receiver to see what he is seeing. It also means that upon reloading, the timeline will be closed.
* unit tests on the query model, query (payload/schema) representers

File: db/migrate/20170404110156_extend_query_model.rb (hunk 3 above)
Comment (oliverguenther) [reply db/migrate/20170404110156_extend_query_model.rb:3]: We'll likely have more timeline related properties (zoom level and whatnot) soon so I don't think thats a good idea.

**Comment by oliverguenther** [comment]
I've added more tests and the timeline attribute is now part of the query_props URL.

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/b7a63280ab94431b6c70b600747d8703c0d6a794) (deployed commit: b6555be667177770bb645be8b97cad919aa39fc1).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/143ce288cbd3228a7651fa7dea91c2662b8cc467) (deployed commit: 931ced89347c72caa891b044e0a57e3576ecb486).

**Comment by oliverguenther** [comment]
I've extracted the unrelated failing spec into https://github.com/opf/openproject/pull/5327

**Review by ulferts — APPROVED** [review]

**Pull request merged**